File: problems/python/198.house-robber.py
from typing import List


class Solution:
    # 2, 7, 9, 3, 1
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        result = 0
        dp = [0] * 100
        dp[0] = nums[0]
        dp[1] = nums[1]

        for house, money in enumerate(nums[2:], start=2):
            max_stole = 0
            for adj in range(house - 2, -1, -1):
                max_stole = max(max_stole, dp[adj])
            dp[house] = max_stole + money

        for num in dp:
            result = max(result, num)
        return result

File: c/198.house-robber.c
#include <stdio.h>
#include <stdlib.h>
#define MAX(a, b) (((a) > (b)) ? (a) : (b))
int rob(int* nums, int numsSize) {
    if (numsSize == 1)
        return nums[0];
    int *dp = (int*)malloc(numsSize * sizeof(int));
    int result = 0;
    dp[0] = nums[0];
    dp[1] = nums[1];

    for (int house = 2; house < numsSize; house++) {
        int max_stole = 0;
        for (int adj = house-2; adj >= 0; adj--) {
            max_stole = MAX(max_stole, dp[adj]);
        }
        dp[house] = max_stole + nums[house];
        printf("%d\n", dp[house]);
    }

    for (int i = 0; i < numsSize; i++) {
        result = MAX(result, nums[i]);
    }

    free(dp);
    return result;
}

int main() {
    int num[] = {2,7,9,3,1};
    int r = rob(num, 5);
}
